File: toolkit/components/uniffi-bindgen-gecko-js/fixtures/tests/xpcshell/test_geometry.js
/* Any copyright is dedicated to the Public Domain.
   http://creativecommons.org/publicdomain/zero/1.0/ */

const Geometry = ChromeUtils.importESModule(
  "resource://gre/modules/RustGeometry.sys.mjs"
);

add_task(async function () {
  const ln1 = new Geometry.Line(
    new Geometry.Point({ coord_x: 0, coord_y: 0 }),
    new Geometry.Point({ coord_x: 1, coord_y: 2 })
  );
  const ln2 = new Geometry.Line(
    new Geometry.Point({ coord_x: 1, coord_y: 1 }),
    new Geometry.Point({ coord_x: 2, coord_y: 2 })
  );
  const origin = new Geometry.Point({ coord_x: 0, coord_y: 0 });
  Assert.ok(
    (await Geometry.intersection({ start: ln1, end: ln2 })).equals(origin)
  );
  Assert.deepEqual(
    await Geometry.intersection({ start: ln1, end: ln2 }),
    origin
  );
  Assert.strictEqual(
    await Geometry.intersection({ start: ln1, end: ln1 }),
    null
  );
});

File: toolkit/components/uniffi-fixtures/geometry/tests/bindings/test_geometry.py
from geometry import Line, Point, gradient, intersection

ln1 = Line(start=Point(coord_x=0, coord_y=0), end=Point(coord_x=1, coord_y=2))
ln2 = Line(start=Point(coord_x=1, coord_y=1), end=Point(coord_x=2, coord_y=2))

assert gradient(ln1) == 2
assert gradient(ln2) == 1

assert intersection(ln1, ln2) == Point(coord_x=0, coord_y=0)
assert intersection(ln1, ln1) is None
